Print partial fx graph on dispatch_trace failures

voznesenskym opened the pull request:
Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89732

voznesenskym committed:
Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.

[ghstack-poisoned]

voznesenskym committed:
Update on "Print partial fx graph on dispatch_trace failures"

Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.




[ghstack-poisoned]

voznesenskym committed:
Update on "Print partial fx graph on dispatch_trace failures"

Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.




[ghstack-poisoned]

voznesenskym committed:
Update on "Print partial fx graph on dispatch_trace failures"

Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.




[ghstack-poisoned]

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -433,13 +436,30 @@ def create_arg(self, a: Any):
+    except Exception:
+        if DISPATCH_TRACE_ERROR_PRINT_GRAPH:
+            gm_printed = GraphModule(tracer.root, tracer.graph, name).print_readable(print_output=False)
+            log.info(f"""
```
info severity is too low, by default our log level is warning not info, so this won't actually show up under the instructions you gave below. If you want to keep the log level this way, then you should probably gate the logging on log.isEnabledFor(INFO), unconditionally print it, and then ask the user to crank up log level. Please... test if the API is working?

But I prefer a specialized enable trace flag, because if I turn on INFO I'm likely to get a bunch of other spew I don't really care about, but if I'm following up on the breadcrumb I saw below, I probably only care about this output. I suppose we can compromise on the earlier code review comment: the breadcrumb should be a log, while the actual triggered print should just be a print.

ezyang replied:
I will not give you any trouble if you follow the functorch convention (although I think functorch is bad for spewing the graphs to stdout, fwiw.)

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -433,13 +436,30 @@ def create_arg(self, a: Any):
+            Dispatch trace exception,
+            set torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH = True
+            to print the incomplete graph.
+            """)
```
The indentation on the string is going to look super ugly when printed. Please dedent it to the left.

ezyang replied:
If you really care about maintaining the relative indent, use string juxtaposition
```
log.warning(
    "Foo "
    "bar "
    "baz."
)
```

voznesenskym committed:
Update on "Print partial fx graph on dispatch_trace failures"

Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.




[ghstack-poisoned]

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -449,16 +449,15 @@ def dispatch_trace(
-            """)
+Dispatch trace exception, set torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH = True
+to print the incomplete graph.
+""")
```
nyeh, you should concatenate the strings together, so the log message is single line. (And even if you didn't, you should remove the leading and trailing newlines)

voznesenskym committed:
Update on "Print partial fx graph on dispatch_trace failures"

Add support for setting torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH to enable users to print partial fx graphs when their dispatch_trace fails. Also wrap the error to make it more clear that it is coming from dispatch trace.




[ghstack-poisoned]

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -433,13 +436,27 @@ def create_arg(self, a: Any):
+""")
+        else:
+            log.warning("""Dispatch trace exception, set torch.fx.experimental.proxy_tensor.DISPATCH_TRACE_ERROR_PRINT_GRAPH = True
+to print the incomplete graph.""")
```
you still have an unnecessary newline in the log message...

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot rebase
